### ISSUE TITLE
序列标注模型蒸馏问题

### ISSUE THREAD

**Issue by OliveLv** [PaddlePaddle/PaddleSlim#1777]
1、crf的模型蒸馏该怎么写？看文心序列标注的蒸馏，主要是fc进行softmax，但crf模型fc后面加了一个crf decode，这种情况也是对teacher模型的fc进行softmax么
2、模型蒸馏step1是直接看loss么，选择loss较低的模型就行？
3、目前随机选择了一个step1 loss较低的，然后进行step2蒸馏，最终模型比teacher模型f1少了0.1+（teacher：0.910143，student：0.77），这种情况正常么？比较怀疑是蒸馏loss写得不正确
4、直接用蒸馏的模型热启，进行finetune，效果和模型蒸馏差不多甚至更好。是不是模型蒸馏写得不对导致的？
模型蒸馏相关代码

teacher_logits/student_logits = fluid.layers.fc(
            size=self._flags.num_out_dimension,
            input=input_feature,
            param_attr=fluid.ParamAttr(
                name=ernie_model._name+'fc_1',
                initializer=fluid.initializer.Uniform(
                    low=-init_bound, high=init_bound),
                regularizer=fluid.regularizer.L2DecayRegularizer(
                    regularization_coeff=1e-4)))

teacher_pred = fluid.layers.crf_decoding(
            input=teacher_logits, param_attr=fluid.ParamAttr(name='crfw'))
## pred_loss
pred_loss = fluid.layers.mean(
                fluid.layers.linear_chain_crf(
                    input=student_logits,
                    label=teacher_pred,
                    param_attr=fluid.ParamAttr(
                        name='crfw',
                        learning_rate=crf_lr)))
## teacher predict loss
teacher_ce_loss = fluid.layers.mean(
                fluid.layers.linear_chain_crf(
                    input=teacher_logits,
                    label=unpad_labels,
                    param_attr=fluid.ParamAttr(
                        name='crfw',
                        learning_rate=crf_lr)))
 ## student predict loss
 student_ce_loss = fluid.layers.mean(
            fluid.layers.linear_chain_crf(
                input=student_logits,
                label=unpad_labels,
                param_attr=fluid.ParamAttr(
                    name='crfw',
                    learning_rate=crf_lr)))

**Comment by wanghaoshuang** [comment]
>1、crf的模型蒸馏该怎么写？看文心序列标注的蒸馏，主要是fc进行softmax，但crf模型fc后面加了一个crf decode，这种情况也是对teacher模型的fc进行softmax么

具体在哪层后边加蒸馏，取决于任务，没有标准。如果 crf decode 中有可学习参数，建议将蒸馏 loss 加到 crf decode 之后。蒸馏 loss也不一定非要 softmax.

>2、模型蒸馏step1是直接看loss么，选择loss较低的模型就行？

理论上是这样。

>3、目前随机选择了一个step1 loss较低的，然后进行step2蒸馏，最终模型比teacher模型f1少了0.1+（teacher：0.910143，student：0.77），这种情况正常么？比较怀疑是蒸馏loss写得不正确

正常。不能拿 student 和 teacher 的精度比，下边4中的对比更合理，也就是和不加蒸馏的 student 比效果。

>4、直接用蒸馏的模型热启，进行finetune，效果和模型蒸馏差不多甚至更好。是不是模型蒸馏写得不对导致的？
模型蒸馏相关代码

建议调整蒸馏策略试试：
- 调整温度系数
- 调整蒸馏 loss 的位置
